accumulator/batchproof.go

```diff
@@ -112,122 +112,134 @@ func FromBytesBatchProof(b []byte) (BatchProof, error) {
 
-			if verbose {
-				fmt.Printf("left %d rootPoss %d\n", left, rootPositions[0])
-			}
-			// check for roots
-			if left == rootPositions[0] {
-				if verbose {
-					fmt.Printf("one left in tagrow; should be root\n")
-				}
-				// Grab the hash of this position from the map
-				computedRootHash, ok := proofmap[left]
-				if !ok {
-					fmt.Printf("ERR no proofmap for root at %d\n", left)
-					return false, nil
-				}
-				// Verify that this root hash matches the one we stored
-				if computedRootHash != roots[0] {
-					fmt.Printf("row %d root, pos %d expect %04x got %04x\n",
-						row, left, roots[0][:4], computedRootHash[:4])
-					return false, nil
-				}
-				// otherwise OK and pop of the root
-				roots = roots[1:]
-				rootPositions = rootPositions[1:]
-				rootRows = rootRows[1:]
-				break
-			}
+		// target should have its sibling in targetNodes
+		if len(targetNodes) == 1 {
+			// sibling not found
+			return false, nil, nil
+		}
 
-			// Grab the parent position of the leaf we've verified
-			parentPos := parent(left, rows)
-			if verbose {
-				fmt.Printf("%d %04x %d %04x -> %d\n",
-					left, proofmap[left], right, proofmap[right], parentPos)
-			}
+		proof = targetNodes[1]
+		targetNodes = targetNodes[2:]
 
-			// this will crash if either is 0000
-			// reconstruct the next row and add the parent to the map
-			parhash := parentHash(proofmap[left], proofmap[right])
-			nextRow = append(nextRow, parentPos)
-			proofmap[parentPos] = parhash
+	hash:
+		// figure out which node is left and which is right
+		left := target
+		right := proof
+		if target.Pos&1 == 1 {
+			right, left = left, right
 		}
 
-		// Make the nextRow the tagRow so we'll be iterating over it
-		// reset th nextRow
-		tagRow = nextRow
-		nextRow = []uint64{}
+		// get the hash of the parent from the cache or compute it
+		parentPos := parent(target.Pos, rows)
+		isParentCached, hash := cached(parentPos)
+		if !isParentCached {
+			hash = parentHash(left.Val, right.Val)
+		}
+		trees = append(trees, [3]node{{Val: hash, Pos: parentPos}, left, right})
 
-		// if done with row and there's a root left on this row, remove it
-		if len(rootRows) > 0 && rootRows[0] == row {
-			// bit ugly to do these all separately eh
-			roots = roots[1:]
-			rootPositions = rootPositions[1:]
-			rootRows = rootRows[1:]
+		row := detectRow(parentPos, rows)
+		if numLeaves&(1<<row) > 0 && parentPos == rootPosition(numLeaves, row, rows) {
+			// the parent is a root -> store as candidate, to check against actual roots later.
+			rootCandidates = append(rootCandidates, node{Val: hash, Pos: parentPos})
+			continue
 		}
+		targetNodes = append(targetNodes, node{Val: hash, Pos: parentPos})
+	}
+	if len(rootCandidates) == 0 {
+		// no roots to verify
+		return false, nil, nil
+	}
+	rootMatches := 0
+	for _, root := range roots {
+		if len(rootCandidates) > rootMatches && root == rootCandidates[rootMatches].Val {
+			rootMatches++
+		}
+	}
+	if len(rootCandidates) != rootMatches {
```

Comment:
Since all roots should match in order for the batchProof to be valid, wouldn't it be better to return false in the block at line 232 if any ```root``` doesn't match a ```rootCandidate```?

Reply:
`rootCandidates` only holds a subset of the roots, so it could happen that a root does not match a candidate but by making sure that all candidates were matched to some root it still works out.
also added a comment there.